### PR TITLE
Linkedin api properly implemented, pagination on the way

### PR DESCRIPTION
**Pull request by Antonhansel**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,27 @@
-# Provider boilerplate
+# Linkedin provider
 > Visit http://anyfetch.com for details about AnyFetch.
 
-Git clone this repo for boilerplate code on your provider.
-Read the documentation from the library on https://github.com/AnyFetch/anyfetch-provider.js.
+AnyFetch provider for files stored in Linkedin
```
Comment (Antonhansel) [inline README.md]: test

Comment (Neamar) [inline README.md]: people

File: README.md [hunk 2]
```diff
@@ -1,7 +1,27 @@
...
+
+You will need to populate your local .env file with the following variables
```
Comment (Neamar) [inline README.md:6]: `.env`

File: README.md [hunk 3]
```diff
@@ -1,7 +1,27 @@
...
+
+```bash
+# Go to https://www.linkedin.com/secure/developer to ask for app id and secret
+export LINKEDIN_API_KEY="linkedin-api"
```
Comment (Neamar) [inline README.md]: no export

File: README.md [hunk 4]
```diff
@@ -1,7 +1,27 @@
...
+export LINKEDIN_SECRET_TOKEN="linkedin-secret"
+
+# Provider URL, most probably https://your-host
+export PROVIDER_URL="https://your-host"
+
+# AnyFetch app id and secret
+export ANYFETCH_API_ID="anyfetch-app-id"
+export ANYFETCH_API_SECRET="anyfetch-app-secret"
+
```
Comment (Neamar) [inline README.md]: also Linkedin fake token?

File: lib/helpers/upload.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+var rarity = require('rarity');
+/**
```
Comment (Neamar) [inline lib/helpers/upload.js:6]: \n\n

File: lib/helpers/upload.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * Upload `contact` (containing contact data) onto AnyFetch.
+ *
+ *
+ * @param {Object} contact Contact to upload, plus anyfetchClient
+ * @param {Object} anyfetchClient Client for upload
+ * @param {Object} accessToken Access token of the current account
+ * @param {Object} contact Contact to upload, plus anyfetchClient
+ * @param {Function} cb Callback to call once contacts has been uploaded.
+ */
+
```
Comment (Neamar) [inline lib/helpers/upload.js]: no \n

File: lib/helpers/upload.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        image: contact.pictureUrl || '',
+        job: contact.headline || '',
+        phone: '',
+        address: contact.location ? contact.location.name : '',
+        website: '',
+        data: {},
+      }
+    };
+
+  delete contact.metadata.siteStandardProfileRequest;
```
Comment (Neamar) [inline lib/helpers/upload.js]: why so?

File: lib/helpers/upload.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * @param {Object} accessToken Access token of the current account
+ * @param {Object} contact Contact to upload, plus anyfetchClient
+ * @param {Function} cb Callback to call once contacts has been uploaded.
+ */
+
+
+module.exports = function upload(contact, anyfetchClient, accessToken, cb) {
+
+  var today = new Date();
+  delete contact.apiStandardProfileRequest;
```
Comment (Neamar) [inline lib/helpers/upload.js]: why so?

File: lib/helpers/upload.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  contact = {
+      identifier: contact.siteStandardProfileRequest.url.split('&')[0], // extract the relation url
+      creation_date: today,
```
Comment (Neamar) [inline lib/helpers/upload.js]: you can leave empty

File: lib/helpers/upload.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      modification_date: today,
```
Comment (Neamar) [inline lib/helpers/upload.js]: yyou can leave empty

File: lib/helpers/upload.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+
+module.exports = function upload(contact, anyfetchClient, accessToken, cb) {
+
+  var today = new Date();
+  delete contact.apiStandardProfileRequest;
+
+  contact = {
+      identifier: contact.siteStandardProfileRequest.url.split('&')[0], // extract the relation url
```
Comment (Neamar) [inline lib/helpers/upload.js]: compute it once above and reuse

File: lib/helpers/upload.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      document_type: 'contact',
+      actions: {
+        'show': contact.siteStandardProfileRequest.url.split('&')[0],
+      },
+      user_access: [anyfetchClient.accessToken],
+      metadata: {
+        name: (contact.firstName || ' ') + ' ' + (contact.lastName || ' '),
+        image: contact.pictureUrl || '',
+        job: contact.headline || '',
+        phone: '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: leave empty

File: lib/helpers/upload.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        address: contact.location ? contact.location.name : '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: address: contact.location ? [contact.location.name] : null,

File: lib/helpers/upload.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        website: '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: empty

File: lib/helpers/upload.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        data: {},
```
Comment (Neamar) [inline lib/helpers/upload.js]: not here

File: lib/index.js [hunk 16]
```diff
@@ -2,17 +2,58 @@
...
+var state = crypto.randomBytes(10).toString('hex');
+
 
 var redirectToService = function(callbackUrl, cb) {
   // Redirect user to provider consentment page
-  cb(null, config.providerUrl + '/init/callback', {some: 'data'});
+  var redirectUrl = 'https://www.linkedin.com/uas/oauth2/authorization?response_type=code&client_id=' + config.linkedin.api +
+  '&state=' + state +
+  '&redirect_uri=' + config.providerUrl + '/init/callback';
+  cb(null, redirectUrl, {redirectUrl: redirectUrl, callbackUrl: callbackUrl});
```
Comment (Neamar) [inline lib/index.js]: maybe redirectUrl is not useful

File: lib/index.js [hunk 17]
```diff
@@ -2,17 +2,58 @@
...
+    return cb(new CancelError());
+  }
+
+  var token;
+  var apiUrl = 'https://api.linkedin.com/';
+
+  async.waterfall([
+    function getToken(cb) {
+      var app = 'https://www.linkedin.com/';
+      var url = '/uas/oauth2/accessToken?grant_type=' + 'authorization_code' +
```
Comment (Neamar) [inline lib/index.js]: ' + '

File: lib/index.js [hunk 18]
```diff
@@ -2,17 +2,58 @@
...
+    function getEmailAddress(res, cb) {
+      token = res.body.access_token;
+      var url = '/v1/people/~/email-address?oauth2_access_token=' + token + '&format=json';
+      request(apiUrl)
+        .get(url)
+        .expect(200)
+        .end(cb);
+    },
+    function callFinalCb(data, cb) {
+      cb(null, data.body, {access_token: token, callbackUrl: storedParams.callbackUrl});
```
Comment (Neamar) [inline lib/index.js]: why callbackUrl?

File: lib/index.js (hunk 16 above)
Comment (Neamar) [inline lib/index.js]: and callbackUrl

File: lib/workers.js [hunk 19]
```diff
@@ -1,15 +1,21 @@
...
+
+var uploadContact = require('./helpers/upload.js');
+
 module.exports.addition = function additionQueueWorker(job, cb) {
-  // Send data to AnyFetch.
-  // You may define this as an helper function
-  // Job is an item with your task, an anyfetch client, and data stored by retrieveTokens()
-  cb(null);
+  log.info({
+		name: 'addition',
```
Comment (Neamar) [inline lib/workers.js]: looks you're using tabs

File: lib/workers.js [hunk 20]
```diff
@@ -1,15 +1,21 @@
...
+		identifier: job.task.identifier
+	}, "Uploading");
+  uploadContact(job.task, job.anyfetchClient, job.serviceData.access_token, cb);
 };
 
 module.exports.deletion = function deletionQueueWorker(job, cb) {
```
Comment (Neamar) [inline lib/workers.js:15]: you can remove it completely

File: package.json [hunk 21]
```diff
@@ -21,10 +21,12 @@
     "PROVIDER"
   ],
   "dependencies": {
-    "async": "0.9.x",
-    "request": "2.49.x",
     "anyfetch-provider": "^3.0.9",
-    "dotenv": "^0.4.0"
+    "async": "0.9.x",
+    "dotenv": "^0.4.0",
+    "node-linkedin": "^0.3.5",
```
Comment (Neamar) [inline package.json]: can be removed

File: test/test.js [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+require('should');
+var async = require('async');
+
+var update = require('../lib/update.js');
+var config = require('../config/configuration.js');
+var uploadContact = require('../lib/helpers/upload.js');
+
+
+describe("Linkedin provider", function() {
+  var connexionsPushed = [];
```
Comment (Neamar) [inline test/test.js]: connectionPushed

File: test/test.js [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+      contact.should.have.property('metadata');
+      count += 1;
+      if(count === connexionsPushed.length) {
+        callback(null);
+      }
+    },
+    accessToken: config.linkedin.fake,
+  };
+
+  it('can list all connections', function(done) {
```
Comment (Neamar) [inline test/test.js]: connexions

File: test/test.js [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+    update({access_token: config.linkedin.fake}, null, fakeQueue, function() {
+      connexionsPushed.length.should.equal(1);
```
Comment (Neamar) [inline test/test.js]: should.be.greaterThan

File: test/test.js (hunk 23 above)
Comment (Neamar) [inline test/test.js]: should list

File: test/test.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+      done();
+    });
+  });
+
+  it('can list only updated contacts', function(done) {
```
Comment (Neamar) [inline test/test.js]: should

File: test/test.js [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    update({access_token: config.linkedin.fake}, new Date(), fakeQueue, function() {
+      connexionsPushed.length.should.equal(1);
```
Comment (Neamar) [inline test/test.js]: eql(0)

File: test/test.js [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+  it('can list all connections', function(done) {
+    update({access_token: config.linkedin.fake}, null, fakeQueue, function() {
+      connexionsPushed.length.should.equal(1);
+      done();
+    });
+  });
+
+  it('can list only updated contacts', function(done) {
+    update({access_token: config.linkedin.fake}, new Date(), fakeQueue, function() {
```
Comment (Neamar) [inline test/test.js]: probably new Date().getTime()

File: test/test.js [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+      connexionsPushed.length.should.equal(1);
+      done();
+    });
+  });
+
+  it('can upload contacts', function(done) {
```
Comment (Neamar) [inline test/test.js]: should

File: test/test.js [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+    async.map(connexionsPushed, function(contact) {
+      uploadContact(contact, fakeClient, config.linkedin.fake, function(err) {
```
Comment (Neamar) [inline test/test.js]: , done)

File: README.md [hunk 30]
```diff
@@ -1,7 +1,49 @@
...
+Register your app on manager-staging.anyfetch.com, start the server in bin/server, fire up a Redis and MongoDB server and visit this url:
+
+https://manager-staging.anyfetch.com/oauth/authorize?client_id=YOUR_CLIENT_ID&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Finit%2Fconnect
+
+You can get a client id under "Client Id" on this page https://manager-staging.anyfetch.com/developers
+
+After your Linkedin account has been linked to the provider, connect to mongo and find your access_token in the tokens collection under 'data'
+
+You can now add it to your .env file and fire up ```npm test```
```
Comment (Neamar) [inline README.md]: `npm test` (only one backtick)

File: README.md [hunk 31]
```diff
@@ -1,7 +1,49 @@
...
+
+# How does it work?
+## Init
+AnyFetch API will call `/init/connect` with anyfetch authorization code. We will generate a request_token and transparently redirect the user to Linkedin consentment page.
+Linkedin will then call us back on `/init/callback`. We'll check our request_token has been granted approval, and store this.
+
+## Update
+To update a users document list curl on localhost:8000
+
+http://localhost:8000/update?access_token=LINKEDIN_ACCESS_TOKEN&documents_per_update=NUMBER_OF_DOCUMENTS&force=1
```
Comment (Neamar) [inline README.md]: it's not the LINKEDIN ACCESS TOKEN, rather the anyfetch_access_token

File: lib/helpers/upload.js [hunk 32]
```diff
@@ -0,0 +1,38 @@
...
+      identifier: identifier,
+      creation_date: '',
+      modification_date: '',
+      document_type: 'contact',
+      actions: {
+        'show': identifier,
+      },
+      user_access: [anyfetchClient.accessToken],
+      metadata: {
+        name: (contact.firstName || ' ') + ' ' + (contact.lastName || ' '),
```
Comment (Neamar) [inline lib/helpers/upload.js]: `(contact.firstName || '') + ' ' + (contact.lastName || '')`

File: lib/helpers/upload.js [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+        image: contact.pictureUrl || '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: you don't need the ||

File: lib/helpers/upload.js [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+        job: contact.headline || '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: you don't need the ||

File: lib/helpers/upload.js (hunk 12 above)
Comment (Neamar) [inline lib/helpers/upload.js]: (remove the line)

File: lib/helpers/upload.js [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+        phone: '',
+        address: contact.location ? [contact.location.name] : [''],
+        website: '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: remove the line

File: lib/index.js [hunk 36]
```diff
@@ -2,17 +2,58 @@
...
+var state = crypto.randomBytes(10).toString('hex');
+
 
 var redirectToService = function(callbackUrl, cb) {
   // Redirect user to provider consentment page
-  cb(null, config.providerUrl + '/init/callback', {some: 'data'});
+  var redirectUrl = 'https://www.linkedin.com/uas/oauth2/authorization?response_type=code&client_id=' + config.linkedin.api +
+  '&state=' + state +
+  '&redirect_uri=' + config.providerUrl + '/init/callback';
+  cb(null, redirectUrl, {callbackUrl: callbackUrl});
```
Comment (Neamar) [inline lib/index.js]: can we remove callbackUrl

File: lib/update.js [hunk 37]
```diff
@@ -1,9 +1,33 @@
...
+
+  async.waterfall([
+   function getConnections(cb) {
+    request('https://api.linkedin.com')
+    .get(url)
+    .expect(200)
+    .end(cb);
+  },
+  function parseConnections(res, cb) {
+    if(res.body.values && res.body.values.length > 0) {
```
Comment (Neamar) [inline lib/update.js]: you don't need to check for the length, in the worst case the forEach will iterate on empty

File: README.md [hunk 38]
```diff
@@ -1,7 +1,49 @@
...
+LINKEDIN_API_KEY="linkedin-api"
+LINKEDIN_SECRET_TOKEN="linkedin-secret"
+
+# Provider URL, most probably https://your-host
+PROVIDER_URL="https://your-host"
+-
+# AnyFetch app id and secret
+ANYFETCH_API_ID="anyfetch-app-id"
+ANYFETCH_API_SECRET="anyfetch-app-secret"
+LINKEDIN_FAKE_ACCESS_TOKEN="Linkedin fake access token" // FOR TESTING PURPOSES
```
Comment (Neamar) [inline README.md]: // => #

File: config/configuration.js [hunk 39]
```diff
@@ -33,6 +32,11 @@ module.exports = {
   providerUrl: process.env.PROVIDER_URL,
   appId: process.env.ANYFETCH_API_ID,
   appSecret: process.env.ANYFETCH_API_SECRET,
+  linkedin: {
```
Comment (Neamar) [inline config/configuration.js]: linkedIn

File: lib/helpers/upload.js [hunk 40]
```diff
@@ -0,0 +1,34 @@
...
+  contact = {
+      identifier: identifier,
+      document_type: 'contact',
+      actions: {
+        'show': identifier,
+      },
+      user_access: [anyfetchClient.accessToken],
+      metadata: {
+        name: (contact.firstName || '') + ' ' + (contact.lastName || ''),
+        image: contact.pictureUrl ? contact.pictureUrl :  '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: image: contact.pictureUrl

File: lib/helpers/upload.js [hunk 41]
```diff
@@ -0,0 +1,34 @@
...
+        job: contact.headline ? contact.headline : '',
```
Comment (Neamar) [inline lib/helpers/upload.js]: job: contact.headline

File: lib/helpers/upload.js [hunk 42]
```diff
@@ -0,0 +1,34 @@
...
+  var identifier = contact.siteStandardProfileRequest.url.split('&')[0]; // extract the relation url
+  contact = {
+      identifier: identifier,
+      document_type: 'contact',
+      actions: {
+        'show': identifier,
+      },
+      user_access: [anyfetchClient.accessToken],
+      metadata: {
+        name: (contact.firstName || '') + ' ' + (contact.lastName || ''),
```
Comment (Neamar) [inline lib/helpers/upload.js]: contact.firstName + ' ' + contact.lastName

File: lib/update.js [hunk 43]
```diff
@@ -1,9 +1,33 @@
...
-  // Update documents from provider
-  // You may define this as an helper function
-  // this function is pinged on update, with the data stored by retrieveTokens()
-  // You can do queues.worker_name.push(task)
-  cb(null, new Date());
+
+  var url = '/v1/people/~/connections?oauth2_access_token=' + serviceData.access_token + '&format=json';
+
+  if(cursor !== null) {
+    url += '&modified=new&modified-since=' + cursor;
```
Comment (Neamar) [inline lib/update.js:12]: Can you add a link?

**Pull request merged**